src/plugins/dashboard/public/application/utils/create_dashboard_app_state.tsx

```diff
@@ -0,0 +1,161 @@
+  DashboardServices,
+} from '../../types';
+import { ViewMode } from '../../embeddable_plugin';
+import { getDashboardIdFromUrl } from '../lib';
+import { syncQueryStateWithUrl } from '../../../../data/public';
+
+const APP_STATE_STORAGE_KEY = '_a';
+
+interface Arguments {
+  osdUrlStateStorage: IOsdUrlStateStorage;
+  stateDefaults: DashboardAppState;
+  services: DashboardServices;
+  instance: any;
+}
+
+export const createDashboardGlobalAndAppState = ({
+  stateDefaults,
+  osdUrlStateStorage,
+  services,
+  instance,
+}: Arguments) => {
+  const urlState = osdUrlStateStorage.get<DashboardAppState>(APP_STATE_STORAGE_KEY);
+  const {
+    opensearchDashboardsVersion,
+    usageCollection,
+    history,
+    data: { query },
+  } = services;
+
+  /* 
+  Function migrateAppState() does two things
+  1. Migrate panel before version 7.3.0 to the 7.3.0 panel structure.
+     There are no changes to the panel structure after version 7.3.0 to the current 
+     OpenSearch version so no need to migrate panels that are version 7.3.0 or higher
+  2. Update the version number on each panel to the current version.
+  */
+  const initialState = migrateAppState(
+    {
+      ...stateDefaults,
+      ...urlState,
+    },
+    opensearchDashboardsVersion,
+    usageCollection
+  );
+
+  const pureTransitions = {
+    set: (state) => (prop, value) => ({ ...state, [prop]: value }),
+    setOption: (state) => (option, value) => ({
+      ...state,
+      options: {
+        ...state.options,
+        [option]: value,
+      },
+    }),
+    // setDashboard: (state)
+  } as DashboardAppStateTransitions;
+
+  const stateContainer = createStateContainer<DashboardAppState, DashboardAppStateTransitions>(
+    initialState,
+    pureTransitions
+  );
+
+  const { start: startStateSync, stop: stopStateSync } = syncState({
+    storageKey: APP_STATE_STORAGE_KEY,
+    stateContainer: {
+      ...stateContainer,
+      get: () => toUrlState(stateContainer.get()),
+      set: (state: DashboardAppStateInUrl | null) => {
+        // sync state required state container to be able to handle null
+        // overriding set() so it could handle null coming from url
+        if (state) {
+          // Skip this update if current dashboardId in the url is different from what we have in the current instance of state manager
+          // As dashboard is driven by angular at the moment, the destroy cycle happens async,
+          // If the dashboardId has changed it means this instance
+          // is going to be destroyed soon and we shouldn't sync state anymore,
+          // as it could potentially trigger further url updates
+          const currentDashboardIdInUrl = getDashboardIdFromUrl(history.location.pathname);
+          if (currentDashboardIdInUrl !== instance.id) return;
+
+          stateContainer.set({
+            ...stateDefaults,
+            ...state,
+          });
+        } else {
+          // Do nothing in case when state from url is empty,
```

Comment:
if no nothing in else, then why need `else`? 😆

Reply:
This logic is ported over: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/dashboard/public/application/dashboard_state_manager.ts#L202

But I agree, this logic can be much more intentional.

Added a TODO here: 
* https://github.com/opensearch-project/OpenSearch-Dashboards/pull/4540
* And tracked under `Enhancements`: https://github.com/opensearch-project/OpenSearch-Dashboards/issues/3365